libevmasm/PeepholeOptimiser.cpp

```diff
@@ -428,6 +428,47 @@ struct TruthyAnd: SimplePeepholeOptimizerMethod<TruthyAnd>
 	}
 };
 
+// (a == true) -> a and (true == a) -> a
+struct TruthyBooleanComparison: SimplePeepholeOptimizerMethod<TruthyBooleanComparison>{
+	static bool applySimple(
+		AssemblyItem const& _pushOrDupN,
+		AssemblyItem const& _iszero,
+		AssemblyItem const& _iszero1,
+		AssemblyItem const& _pushOrDupN1,
+		AssemblyItem const& _iszero2,
+		AssemblyItem const& _iszero3,
+		AssemblyItem const& _sub,
```

Comment:
Note that this catches only the simplest `a == true` case. Any even slightly more complex expression will slip through. E.g. `(1 + 1 == 2) == true` looks like this:
```
      0x01
      iszero
      iszero
      0x02
      dup1
      eq
      iszero
      iszero
      sub
      tag_11
      jumpi
```

Not sure if we can do anything about that. And maybe it's not worth it. Even `a == true` is a bit of an anti-pattern and I'd expect more complex expressions to be rare.

Reply:
One of the more complex expressions which I could see actually happening is `send()`:
```solidity
    function id(address payable a) external returns (bool) {
        if (a.send(1) == true)
            return true;
        return false;
    }
```
But that looks way to complex for peephole optimizer;
```
      0x01
      iszero
      iszero
      dup3
      0xffffffffffffffffffffffffffffffffffffffff
      and
      0x08fc
      0x01
      swap1
      dup2
      iszero
      mul
      swap1
      mload(0x40)
      0x00
      mload(0x40)
      dup1
      dup4
      sub
      dup2
      dup6
      dup9
      dup9
      call
      swap4
      pop
      pop
      pop
      pop
      iszero
      iszero
      sub
      tag_12
      jumpi
```